packages/extension-driver-redshift/src/lib/redshiftDataSource.ts

```diff
@@ -0,0 +1,184 @@
+      }
+
+      const executeStatementCommand = new ExecuteStatementCommand(executeStatementCommandParams);
+      const statementCommandResult = await redshiftClient.send(executeStatementCommand);
+      return await this.getResultFromExecuteStatement(statementCommandResult, redshiftClient);
+    } catch (e: any) {
+      this.logger.debug(
+        `Errors occurred, release connection from ${profileName}`
+      );
+      redshiftClient.destroy();
+      throw e;
+    }
+  }
+
+  public async prepare({ parameterIndex }: RequestParameter) {
+    // see the section of Running SQL statements with parameters when calling the Amazon Redshift Data API
+    // https://docs.aws.amazon.com/redshift/latest/mgmt/data-api.html
+    return `:${parameterIndex}`;
+  }
+
+  private async testConnection(profileName: string): Promise<DataResult | undefined> {
+    const { redshiftClient, options } = this.redshiftClientMapping.get(profileName)!; 
+    const executeStatementCommandParams: ExecuteStatementCommandInput = {
+      Sql: 'select 1',
+      Database: options!.Database,
+      WorkgroupName: options!.WorkgroupName,
+    };
+
+    const executeStatementCommand = new ExecuteStatementCommand(executeStatementCommandParams);
+
+    try {
+      const statementCommandResult = await redshiftClient.send(executeStatementCommand);
+      return await this.getResultFromExecuteStatement(statementCommandResult, redshiftClient);
+    } catch (e) {
+      redshiftClient.destroy();
+      throw e;
+    }
+  }
+
+  private async getResultFromExecuteStatement(
+    statementCommandResult: ExecuteStatementCommandOutput,
+    redshiftClient: RedshiftDataClient
+  ): Promise<DataResult> {
+    let describeStatementResponse: DescribeStatementResponse | undefined;
+    const describeStatementRequestInput: DescribeStatementCommandInput = {
+      Id: statementCommandResult.Id,
+    };
+
+    // definition of describeStatementResponse.Status
+    // https://github.com/aws/aws-sdk-js-v3/blob/29056f4ca545f7e5cf951b915bb52178305fc305/clients/client-redshift-data/src/models/models_0.ts#L604
+    while (!describeStatementResponse || describeStatementResponse.Status !== 'FINISHED') {
+      const describeStatementCommand = new DescribeStatementCommand(describeStatementRequestInput);
+      describeStatementResponse = await redshiftClient.send(describeStatementCommand);
+
+      if (
+        describeStatementResponse.Status === 'ABORTED' || 
+        describeStatementResponse.Status === 'FAILED'
+      ) {
+        throw describeStatementResponse.Error
+      }
+    }
+
+    const getStatementResultCommandParams: GetStatementResultCommandInput = {
+      "Id": describeStatementResponse.Id
+    };
+    const getStatementResultCommand = new GetStatementResultCommand(getStatementResultCommandParams);
+    const getStatementResultResponse = await redshiftClient.send(getStatementResultCommand);
```

Comment:
It looks like the result might return in batches, we ought to handle the `NextToken` of the response, store the token then fetch the next chunk data when we get request from downstream might be a good idea.

 ```ts
new Readable({
...
  read() {
    if(nextToken) { 
       // GetStatementResultCommand with nextToken
    }
  }
})

Reply:
I've handled the `NextToken` condition at line 155.

Reply:
Thanks, we won't lose data in this way.
If we encounter performance issues in the future, we could consider taking backpressure in Node.js streams into account~